### ISSUE TITLE
startup run-gulp-in-docker.sh serve, dashboard 1.10 not work

### ISSUE THREAD

**Issue by ml1268** [kubernetes/dashboard#3259]
## Environment
-  go   1.11
-  node  v9.11.2
-  npm   6.4.1
-  gulp  3.9.1
-  docker 1.9.1
-  java 1.8.0_181

- Dashboard version : v1.10.0
- Kubernetes version: v1.11
- Operating system: centos7

 ## Steps to reproduce
```

1. kubectl proxy --port 8080 --address "0.0.0.0" --accept-hosts "^*$"
2. git clone https://github.com/kubernetes/dashboard.git
3. sh build/run-gulp-in-docker.sh serve. listening in port 9091
4. I use nginx to proxy the request from localhost.

```
It's looks good, but when I open the chrome window to look dashboard homepage, some error stopped me.
**Look to issue 1.**

I try another way to check what is will happen.
```

1.Follow step 2 above, then I modify the run-gulp-in-docker.sh with 'docker run -itd '.
2.Then I use 'docker exec -it contianer_id /bin/sh' attach to the container(I know this way is incorrect,but it's seem to help me).
3.Next, I use 'gulp serve' to startup the proxy created by gulp, and use 'curl http://localhost:9091/api/v1/services/', nothing but wait timeout.

```
**Something is unreadable. Look to issue 2.**

# issue
common problem:
```
# gulp serve
[08:44:41] **Failed to load external module @babel/register**
[08:44:41] Requiring external module babel-register

```
use "npm install babel/register" unable to work.

I refer to useful experiences from contributors's comments, such 'remove node_modules and npm i' ,but the prompt still exist.

# issue 1.
![image](https://user-images.githubusercontent.com/12624153/45541713-6401ec00-b842-11e8-9d2f-2663d4b58c7b.png)

- /api/appConfig.json request fail
- request other api-server api fail

# issue 2. 
```
[host ~]$ docker exec -it 07b780556af4 /bin/sh
# gulp serve
[09:02:03] Failed to load external module @babel/register
[09:02:03] Requiring external module babel-register
[09:02:05] Using gulpfile /dashboard/gulpfile.babel.js
[09:02:05] Starting 'clean-packaged-backend-source'...
[09:02:05] Starting 'kill-backend'...
[09:02:05] Finished 'kill-backend' after 100 μs
[09:02:05] Starting 'locales-for-backend:dev'...
[09:02:05] Starting 'scripts'...
[09:02:05] Starting 'styles'...
[09:02:05] Starting 'buildExistingI18nCache'...
[09:02:05] Finished 'buildExistingI18nCache' after 226 ms
[09:02:05] Starting 'angular-templates'...
[09:02:05] Finished 'locales-for-backend:dev' after 448 ms
[09:02:05] Finished 'clean-packaged-backend-source' after 485 ms
[09:02:05] Starting 'package-backend-source'...
[09:02:09] Finished 'scripts' after 4.39 s
[09:02:10] Finished 'styles' after 4.66 s
[09:02:10] Starting 'index'...
[09:02:10] Finished 'index' after 66 ms
[09:02:10] Finished 'angular-templates' after 4.57 s
[09:02:10] Starting 'watch'...
[09:02:10] Finished 'watch' after 318 ms
[09:02:10] Finished 'package-backend-source' after 4.79 s
[09:02:10] Starting 'link-vendor'...
[09:02:10] Finished 'link-vendor' after 202 μs
[09:02:10] Starting 'package-backend'...
[09:02:10] Finished 'package-backend' after 15 μs
[09:02:10] Starting 'backend'...
[09:02:10] Finished 'backend' after 391 ms
[09:02:10] Starting 'spawn-backend'...
[09:02:10] Finished 'spawn-backend' after 9.37 ms
[09:02:10] Starting 'serve'...
[HPM] Proxy created: /api  ->  http://localhost:9091
2018/09/14 09:02:10 Starting overwatch
2018/09/14 09:02:10 Using apiserver-host location: http://localhost:8080
2018/09/14 09:02:10 Skipping in-cluster config
2018/09/14 09:02:10 Using random key for csrf signing
2018/09/14 09:02:10 No request provided. Skipping authorization
[09:02:10] Finished 'serve' after 14 ms
2018/09/14 09:02:10 Successful initial request to the apiserver, version: v1.11.1-beta.0.6+c4a121ef15c75f
2018/09/14 09:02:10 Generating JWE encryption key
2018/09/14 09:02:10 New synchronizer has been registered: kubernetes-dashboard-key-holder-kube-system. Starting
2018/09/14 09:02:10 Starting secret synchronizer for kubernetes-dashboard-key-holder in namespace kube-system
[BS] [BrowserSync SPA] Running...
[Browsersync] Access URLs:
 --------------------------------------
       Local: http://localhost:9092/
    External: http://10.10.15.253:9092/
 --------------------------------------
          UI: http://localhost:3002
 UI External: http://10.10.15.253:3002
 --------------------------------------
[Browsersync] Serving files from: /dashboard/.tmp/serve
[Browsersync] Serving files from: /dashboard/src/app
2018/09/14 09:02:11 Initializing JWE encryption key from synchronized object
2018/09/14 09:02:11 Creating in-cluster Heapster client
2018/09/14 09:02:11 Serving insecurely on HTTP port: 9091
2018/09/14 09:02:11 listen tcp 127.0.0.1:9091: bind: address already in use

```

- curl http://localhost:8080/api/v1/nodes success
- curl http://localhost:9091/api/v1/nodes fail
`Error occured while trying to proxy to: localhost:9091/api/v1/nodes`
**curl http://localhost:9091 success,but fail in brower because visit '/api/appConfig.json' fail** 

## Expected result
Help me get it to work
How to debug and look log information?

**Comment by jeefy** [comment]
@ml1268 After you run `sh build/run-gulp-in-docker.sh serve` do you get the following error?
```
[HPM] Proxy created: /api  ->  http://localhost:9091
2018/09/18 03:22:44 Starting overwatch
2018/09/18 03:22:44 Using apiserver-host location: http://localhost:8080
2018/09/18 03:22:44 Skipping in-cluster config
2018/09/18 03:22:44 Using random key for csrf signing
2018/09/18 03:22:44 No request provided. Skipping authorization
2018/09/18 03:22:44 Error while initializing connection to Kubernetes apiserver. This most likely means that the cluster is misconfigured (e.g., it has invalid apiserver certificates or service account's configuration) or the --apiserver-host param points to a server that does not exist. Reason: Get http://localhost:8080/version: dial tcp 127.0.0.1:8080: connect: connection refused
Refer to our FAQ and wiki pages for more information: https://github.com/kubernetes/dashboard/wiki/FAQ
[03:22:44] Finished 'serve' after 30 ms
```

I'm guessing so. This is because Dashboard is starting within a container, and (by default) looking for an API Server at `localhost:8080`. Within the scope of the container, `localhost` won't point to a Kubernetes cluster and causes things to fail.

To get around this, you can run `sh build/run-gulp-in-docker.sh serve --apiserver-host=http://your.working.cluster:port` which tells the container where to find an API server. 

I'm going to close this issue, as it is intended behavior. Feel free to re-open it if I'm wrong.

Thanks!
/remove-kind bug 
/triage support
/close

**Comment by ml1268** [comment]
Yep, I solve this problem by use 'host.docker.internal' DNS name when I connect to host server in docker container which docker engine version is 18.03 for mac. 
I get it work in Tag: 1.8.3  

**host.docker.internal** supported by docker since version 18.03
This is an update from **docker.for.mac.localhost**, available since version 17.06, and **docker.for.mac.host.internal**, available since version 17.12, which may also still work.
